src/index.ts

```diff
@@ -0,0 +1,221 @@
+   */
+  root?: string;
+
+  /**
+   * The path relative to `root` of the entrypoint file that will be served for
+   * app-shell style projects. Usually this is index.html.
+   */
+  entrypoint?: string;
+
+  /**
+   * The path relative to `root` of the app shell element.
+   */
+  shell?: string;
+
+  /**
+   * The path relative to `root` of the lazily loaded fragments. Usually the
+   * pages of an app or other bundles of on-demand resources.
+   */
+  fragments?: string[];
+
+  /**
+   * List of glob patterns, relative to root, of this project's sources to read
+   * from the file system.
+   */
+  sources?: string[];
+
+  /**
+   * List of file paths, relative to the project directory, that should be included
+   * as dependencies in the build target.
+   */
+  dependencies?: string[];
+}
+
+export class ProjectConfig {
+
+  readonly root: string;
+  readonly entrypoint: string;
+  readonly shell: string;
+  readonly fragments: string[];
+  readonly sources: string[];
+  readonly dependencies: string[];
+
+  readonly allFragments: string[];
+
+  /**
+   * Given an absolute file path to a polymer.json-like ProjectOptions object,
+   * read that file. If no file exists, null is returned. If the file exists
+   * but there is a problem reading or parsing it, throw an exception.
+   */
+  static loadOptionsFromFile(filepath: string): ProjectOptions {
+    let configParsed: ProjectOptions;
+    try {
+      const configContent = fs.readFileSync(filepath, 'utf-8');
+      configParsed = JSON.parse(configContent);
+    } catch (error) {
+      // swallow "not found" errors because they are so common / expected
+      if (error.code === 'ENOENT') {
+        logger.debug('no polymer config file found', {file: filepath});
+        return null;
+      }
+      // otherwise, throw an exception
+      throw error;
+    }
+
+    return configParsed;
+  }
+
+  /**
+   * Given an absolute file path to a polymer.json-like ProjectOptions object,
+   * return a new ProjectConfig instance created with those options.
+   */
+  static loadConfigFromFile(filepath: string): ProjectConfig {
+    let configParsed = ProjectConfig.loadOptionsFromFile(filepath);
+    if (!configParsed) {
+      return null;
+    }
+    return new ProjectConfig(configParsed);
+  }
+
+  /**
+   * constructor - given a ProjectOptions object, create the correct project
+   * configuration for those options. This involves setting the correct
+   * defaults, validating options, warning on deprecated options, and
+   * calculating some additional properties.
+   */
+  constructor(options: ProjectOptions) {
+    options = options || {};
+    options = fixDeprecatedOptions(options);
+
+    /**
+     * root
+     */
+    if (options.root) {
+      this.root = path.resolve(options.root);
+    } else {
+      this.root = process.cwd();
+    }
+
+    /**
+     * entrypoint
+     */
+    if (options.entrypoint) {
+      this.entrypoint = path.resolve(this.root, options.entrypoint);
```

Comment:
fyi, the reason why I like ternaries isn't because they're shorter, but because conceptually there's a single assignment, it's just a decision on what the value is. I often write ternaries like:

```javascript
this.foo = (bar == baz)
  ? oneThing
  : orAnother;
```